core/typeSugar.ml

```diff
@@ -4139,11 +4147,39 @@ let rec type_check : context -> phrase -> phrase * Types.datatype * Usage.t =
                    let body = type_check (henv ++ pattern_env pat) body in
                    let () = unify ~handle:Gripers.handle_branches
                           (pos_and_typ body, no_pos bt) in
+                   (* see the comments in eff_cases for the meaning of vs and vs' *)
                    let vs = Env.domain (pattern_env pat) in
-                   let vs' = Env.domain henv.var_env in
+                   let vs' = Env.domain <| List.fold_left (fun env p -> Env.extend env (pattern_env p))
+                                           Env.empty (List.map fst params)
+                   in
                    let us =
                      let vs'' = Ident.Set.union vs vs' in
                      Usage.restrict (usages body) vs''
+                   in
+                    (* check the usages of linear parameters in handler clauses *)
+                    let () =
+                    Env.iter (fun v t ->
+                      let uses = Usage.uses_of v (usages body) in
+                        if uses <> 1 then
+                          if Types.Unl.can_type_be t then
+                            Types.Unl.make_type t
+                          else
+                            Gripers.non_linearity pos uses v t)
+                      (pattern_env pat)
+                   in
+                   (* check the usages of environment linear variables in deep handlers *)
+                   let () =
+                    if descr.shd_depth = Deep then
+                      Usage.iter
+                        (fun v _ ->
+                          if not (StringSet.mem v vs) then
+                            let t = Env.find v henv.var_env in
+                            if Types.Unl.can_type_be t then
+                              Types.Unl.make_type t
+                            else
+                              Gripers.die pos ("Variable " ^ v ^ " of linear type " ^ Types.string_of_datatype t ^ " is used in a deep handler."))
```

Comment:
You can define a `Griper` for this message. See the `Gripers` code near the top of the file.